src/main/java/com/endava/internship/service/UserServiceImpl.java

```diff
@@ -2,68 +2,105 @@
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .filter(user -> user.getPrivileges().contains(Privilege.UPDATE) && user.getAge() > age)
+                .findAny();
     }
 
     @Override
     public Map<Integer, List<User>> groupByCountOfPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .collect(Collectors.groupingBy(user -> user.getPrivileges().size()));
     }
 
     @Override
     public double getAverageAgeForUsers(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .mapToDouble(User::getAge)
+                .average().orElse(-1);
     }
 
     @Override
     public Optional<String> getMostFrequentLastName(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        final var map = getNumberOfLastNames(users);
+        final var list = map.entrySet().stream()
+                .filter(stringLongEntry -> stringLongEntry.getValue().equals(map.values().stream().max(Long::compare).get()))
+                .map(Map.Entry::getKey)
+                .collect(Collectors.toList());
+        return list.size() > 1 ? Optional.empty() : list.stream().findFirst();
     }
 
     @Override
     public List<User> filterBy(final List<User> users, final Predicate<User>... predicates) {
-        throw new UnsupportedOperationException("Not implemented");
+        if (predicates.length < 1) {
+            return users;
+        }
+        var stream = users.stream();
+        for (Predicate<User> predicate : predicates) {
+            stream = stream.filter(predicate);
+        }
+        return stream.collect(Collectors.toList());
     }
 
     @Override
     public String convertTo(final List<User> users, final String delimiter, final Function<User, String> mapFun) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .map(mapFun)
+                .collect(Collectors.joining(","));
     }
 
     @Override
-    public Map<Privilege, List<User>> groupByPrivileges(List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+    public Map<Privilege, List<User>> groupByPrivileges(final List<User> users) {
+        return Arrays.stream(Privilege.values())
+                .collect(Collectors.toMap(privilege -> privilege,
+                        privilege -> users.stream()
+                                .filter(user -> user.getPrivileges().contains(privilege))
+                                .collect(Collectors.toList())));
+
     }
 
     @Override
     public Map<String, Long> getNumberOfLastNames(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
```

Comment:
good, but it can be upgraded to one .collect method
without your rmap

Reply:
Done